src/main/java/com/google/codeu/servlets/FormHandlerServlet.java

```diff
@@ -0,0 +1,76 @@
+package com.google.codeu.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.IOException;
+import java.util.List;
+
+import java.util.Map;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * When the user submits the form, Blobstore processes the file upload
+ * and then forwards the request to this servlet. This servlet can then
+ * process the request using the file URL we get from Blobstore.
+ */
+@WebServlet("/upload")
+public class FormHandlerServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(request, "image");
+
+    // Output some HTML that shows the data the user entered.
+    // A real codebase would probably store these in Datastore.
+    ServletOutputStream out = response.getOutputStream();
+    out.println("<p>Here's the image you uploaded:</p>");
+    out.println("<a href=\"" + imageUrl + "\">");
```

Comment:
You might consider using String.format() here instead to clean up all the string concatenation.

Reply:
Hmm, so that would be something like `String.format("<p>blah:</p>\n<a href=\"%1$s\"><img src=\"%1$s\"/></a>", imageUrl);` 

That has the benefit of not repeating the imageUrl and keeping all the html in one line. But upon reflection that might actually be making the code less readable. Feel free to leave it as-is.